Add unit test checking to `cargo check`

ehuss opened the pull request:
This is an extension of PR #4039, fixing #3431, #4003, #4059, #4330.  The fixes for #4059 can potentially be separated into a separate PR, though there may be some overlap.

The general gist of the changes:

- Add `--profile test` flag to `cargo check` that will enable checking of unit tests.
- `--tests` will implicitly enable checking of unit tests within the lib (checks the same targets as `cargo test`).  This affects the `check`, `test`, and `build` commands.  
- `--benches` behaves similarly by using the same targets as `cargo bench`.
- Fix erroneously linking tests when run with `--test`.

There is one thing I did not do because I wanted more feedback on what others think the expected behavior should be.  What should the behavior of `--all-targets` be?  This patch does not (yet) make any changes to its behavior.  My initial thinking is that it should *add* a target of `--lib --bins --profile test`, but that essentially means the lib and bin targets will be checked twice (and thus any errors/warnings outside of `#[cfg(test)]` will appear twice, which may be confusing, and generally take twice as long to run).  I can add that, but I think it would require adding a new `All` variant to `CompileFilter` so that the code in `generate_targets` can detect this scenario.  I wanted feedback before making a more extensive change like that.  The downside of not adding it is that `--all-targets` will ignore unit tests (if you don't specify `--profile test`).

Summary of the profiles used with this patch:

Command                         | Lib               | Bin foo     | Test t1 | Example e1 | Bench b1 |
-------                         | ---               | -------     | ------- | ---------- | -------- |
`check`                         | check             | check       | -       | -          | -        |
`check --profile test`          | check_test†       | check_test† | -       | -          | -        |
`check --lib`                   | check             | -           | -       | -          | -        |
`check --lib --profile test`    | check_test†       | -           | -       | -          | -        |
`check --bin foo`               | check             | check       | -       | -          | -        |
`check -–bin foo –profile test` | check_test†       | check_test† | -       | -          | -        |
`check --bins`                  | check             | check       | -       | -          | -        |
`check --test t1`               | check             | check       | check_test   | -          | -        |
`check --tests`                 | check, check_test†  | check, check_test† | check_test | check†, check_test†  | -    |
`check --all-targets`           | check, check_test†  | check, check_test†  | check_test   | check, check_test† | check_test    |

† = different behavior from today

rust-highfive commented:
r? @matklad

(rust_highfive has picked a reviewer for you, use r? to override)

ehuss commented:
Closing till I fix this.

ehuss commented:
I've discovered some problems with this PR, I will be fixing them soon.

bors commented:
:umbrella: The latest upstream changes (presumably #4616) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented on src/bin/check.rs:
```diff
@@ -114,10 +130,10 @@ pub fn execute(options: Options, config: &mut Config) -> CliResult {
-        filter: ops::CompileFilter::new(options.flag_lib,
-                                        &options.flag_bin, options.flag_bins,
+        filter: ops::CompileFilter::new(options.flag_lib || options.flag_tests,
+                                        &options.flag_bin, options.flag_bins || options.flag_tests,
```
Could you add a comment to why these two `||` conditions were added?

alexcrichton commented on src/bin/check.rs:
```diff
@@ -106,6 +111,17 @@ pub fn execute(options: Options, config: &mut Config) -> CliResult {
                                     &options.flag_exclude,
                                     &options.flag_package)?;
 
+    let test = options.flag_tests ||
```
This logic may cause subtle oddities, although that's maybe ok for now? but for example:

```
cargo check --tests --bin foo
```

I think that'll check the binary `foo` in "test mode" rather than "do you compile mode", right?

ehuss replied:
You're right.  The more I think about it, it is probably wrong.  `check --tests` should not check bins with the test profile.  I'll fix it.

alexcrichton commented on src/cargo/ops/cargo_rustc/context.rs:
```diff
@@ -862,7 +862,7 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
 
         // Integration tests/benchmarks require binaries to be built
-        if unit.profile.test &&
+        if unit.profile.test && !unit.profile.check &&
```
Could you expand the above comment to explain this negation?

ehuss commented:
OK, I have updated it with a different approach.  After thinking about it more, I think people expect `check --tests` to check all tests, similar to `cargo test`.  So I have reworked it so that `--tests` uses the same behavior as `cargo test`.  This means it will also affect other commands that use `--tests` (like `build`, `bench`, and `test`).  This also makes `test --tests` somewhat redundant, but consistent across the commands.  I'm not entirely sure if that's for the best, I'm not sure which behavior is more surprising.

I also updated `--benches` to have similar behavior (mimicking what happens with `cargo bench`).  Finally, `--all-targets` now really does all targets.  I updated the summary above.

Also, I have so far been ignoring doctests.  Since I believe that will rely on rustdoc, it might be outside of the domain of `cargo check`?

alexcrichton commented:
Gah sorry for the delay! In any case that all sounds great to me, thanks so much!

@bors: r+

bors commented:
:pushpin: Commit efc0dd1 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton efc0dd1b070f5ca61bd2641c6c9b5440842bdc36 -->

bors commented:
:hourglass: Testing commit efc0dd1b070f5ca61bd2641c6c9b5440842bdc36 with merge 8be175e37dbd2bcd84078b1a3512c5ac04ae77cb...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.2990), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/294955818?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 8be175e37dbd2bcd84078b1a3512c5ac04ae77cb to master...

ehuss commented:
Thanks @alexcrichton, I know you are super busy (I'm still not convinced you are just one human).  I had some questions:

* Is there anything I can do to help with the release notes process?
* I noticed bors only closed one of the issues.  Should I add comments to each one explaining how it is fixed?

I assume this will get picked up in nightly the next time Cargo is updated in the main rust repo for people to test.  I feel like this is a subtle, but significant change in behavior.  Hopefully most people will be happy with the changes.

alexcrichton commented:
> Is there anything I can do to help with the release notes process?

Certainly! There's a [PR upstream](https://github.com/rust-lang/rust/pull/45454) for the 1.22 release (which this won't be included in), but you can start a new section for the 1.23 release which would be awesome!

>  Should I add comments to each one explaining how it is fixed?

If you're willing, that'd be awesome! If not, feel free to just leave a comment saying it should be closed now.

> I assume this will get picked up in nightly the next time Cargo is updated in the main rust repo for people to test. 

Indeed! If you'd like to accelerate that, feel free to send a PR yourself to rust-lang/rust :)

ehuss commented:
It looks like @Aaronepower has been writing the release notes.  I'll just leave this here for when he gets started on 1.23, and let him format/wordsmith as desired.

```
Cargo
-----
- [`cargo check --profile test` added to check unit tests][cargo/4592]
- [`--tests`, `--benches`, and `--all-targets` flags will now include all
  potential targets for their respective type][cargo/4592]

[cargo/4592]: https://github.com/rust-lang/cargo/pull/4592
```

jonhoo commented:
@ehuss how does `--all-targets` interact with `--all`? Can you specify both?

ehuss commented:
> @ehuss how does --all-targets interact with --all? Can you specify both?

@jonhoo It should work (it should build every target of every package).  Are you experiencing any issues?

jonhoo commented:
Sorry, no, this was a case of asking before trying. Seems to work as expected! It just wasn't documented anywhere that these work together, though I guess there isn't a good reason why they shouldn't. I suspect it would have helped if the docs somewhere said that selecting target crates and selecting profiles are orthogonal actions, but meh.

ehuss commented:
> selecting target crates and selecting profiles are orthogonal actions, but meh.

I've seen that target and profile selection is a common source of confusion.  Along with improving documentation, I wonder if it might be helpful to display each target/profile being built when -v is used (trying to understand that from the rustc command-line isn't exactly easy).
